packages/react-client/src/hooks/useTrackManager.ts

```diff
@@ -18,15 +18,24 @@ export const useTrackManager = ({ mediaManager, tsClient }: TrackManagerConfig):
   const [paused, setPaused] = useState<boolean>(false);
   const [currentTrackMiddleware, setCurrentTrackMiddleware] = useState<TrackMiddleware>(null);
   const type = TRACK_TYPE_TO_DEVICE[mediaManager.getDeviceType()];
+  const joined = useRef<boolean>(false);
```

Comment:
**suggestion:** i think the `usePeerStatus` hook would've come in handy here

Reply:
Unfortunately, the status from `usePeerStatus` could be outdated.

If I understand the React flow correctly, the toggle function remembers the status of the invocation. So, if it starts with null, it's possible that this state changes to joined while the device is starting, but the stream functions will remember the outdated state. Additionally, the toggle function will still have a reference to the outdated stream method.

Reply:
Every function is also recomputed at react render. So if you use a state and capture it in a closure, it won't go out of sync, because after a render, that's a different function with different closure state.

Reply:
https://codesandbox.io/p/sandbox/mt4l88

compare calling those two `setDisplayedMessage` variants, they will behave the same

Reply:
I'm pushing this solution because I want to keep the code here simple.
If we need a ref with `joined` state - it should land in another hook like `useJoinedStatusRef`.
Though I think we don't need that and already created hook should make do here.

Reply:
> `joined.current` is used at the top of the function before any async code, so we can be certain the value hasn't changed between invoking the function and using the value.

Yeah, this value is stable state inside this function.

The problem arises when the user invokes the toggle function. That function saves a reference to the `stream` function, which holds a particular state at the time it was called. Even if the `peerState` changes, it will not be reflected in future `toggle` invocations.

```
const peerStatus = usePeerStatus()
const stream = async () => {
    // peerStatus at the moment of `toggle` invocation 
    if (!joined.current) return; // it does not deopend on memoized value because it mutates

    if (currentTrack?.trackId) {
      await resumeStreaming();
    } else {
      await startStreaming();
    }
 };

  async function toggle(mode: ToggleMode) {
    // ... 
      await mediaManager.start(true); // async
      // peerState can chage here
      await stream(); // reference to function with outdated state, function with new state will be available in next invocation  
  }
```

Reply:
alright, let's add `getCurrentPeerState` method based on a ref in `usePeerState`